python/ray/data/_internal/execution/streaming_executor_state.py

```diff
@@ -181,21 +189,55 @@ def select_operator_to_run(topology: Topology) -> Optional[PhysicalOperator]:
     operators with a large number of running tasks `num_active_tasks()`.
     """
 
-    # TODO: set limits properly based on resources and execution options. This is just
-    # a hard-coded development placeholder.
-    PARALLELISM_LIMIT = 8
-    num_active_tasks = sum(
-        op_state.num_active_tasks() for op_state in topology.values()
-    )
-    if num_active_tasks >= PARALLELISM_LIMIT:
-        return None
+    # Filter to ops that are eligible for execution.
+    ops = [
+        op
+        for op, state in topology.items()
+        if state.num_queued() > 0 and _execution_allowed(op, cur_usage, limits)
+    ]
+
+    # To ensure liveness, allow at least 1 op to run regardless of limits.
+    if not ops and all(op.num_active_work_refs() == 0 for op in topology):
+        # The topology is entirely idle, so choose from all ready ops ignoring limits.
```

Comment:
We'll probably need to revisit this once we have an actual consumer that we want to feed (we can end up overloading / spilling if we ignore the limits when there's a slow consumer).

Reply:
That's a good point. The only thing saving us is that the control loop idles when the consumer is blocked--- but if we move the control loop to a background thread the problem can pop up again.